### PR TITLE
trojan-rs: init at 0.16.0-unstable-2024-11-21

### PR DESCRIPTION
**Pull request by oluceps**
## Description of changes
Add https://github.com/lazytiger/trojan-rs
related https://github.com/NixOS/nixpkgs/issues/287770

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by linsui** [comment]
Try `RUSTC_BOOTSTRAP = true;`.

**Comment by oluceps** [comment]
> Try `RUSTC_BOOTSTRAP = true;`.

I've tried and got the same error. Tried using `fenix` as well

**Comment by linsui** [comment]
I guess this has been fixed in master.

**Comment by oluceps** [comment]
> I guess this has been fixed in master.

Nope.

```
trojan-rs>    Compiling tracing v0.1.37
trojan-rs>    Compiling clap v4.0.32
trojan-rs> error[E0554]: `#![feature]` may not be used on the stable release channel
trojan-rs>   --> /build/trojan-rs-0.16.0-vendor.tar.gz/ipset/src/lib.rs:70:1
trojan-rs>    |
trojan-rs> 70 | #![feature(c_variadic)]
trojan-rs>    | ^^^^^^^^^^^^^^^^^^^^^^^
trojan-rs> 
trojan-rs> error[E0554]: `#![feature]` may not be used on the stable release channel
trojan-rs>   --> /build/trojan-rs-0.16.0-vendor.tar.gz/ipset/src/lib.rs:71:1
trojan-rs>    |
trojan-rs> 71 | #![feature(concat_idents)]
trojan-rs>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
trojan-rs> 
trojan-rs> error[E0554]: `#![feature]` may not be used on the stable release channel
trojan-rs>   --> /build/trojan-rs-0.16.0-vendor.tar.gz/ipset/src/lib.rs:71:12
trojan-rs>    |
trojan-rs> 71 | #![feature(concat_idents)]
trojan-rs>    |            ^^^^^^^^^^^^^
trojan-rs> 
trojan-rs> For more information about this error, try `rustc --explain E0554`.
trojan-rs> error: could not compile `ipset` (lib) due to 3 previous errors
trojan-rs> warning: build failed, waiting for other jobs to finish...
```

**Comment by linsui** [comment]
Yep, nightly rust is required so you need to also add `RUSTC_BOOTSTRAP = true;`. But looks like there are also other problems.

**Comment by linsui** [comment]
`RUSTFLAGS = "--cfg tokio_unstable";` is also needed.

**Comment by oluceps** [comment]
> `RUSTFLAGS = "--cfg tokio_unstable";` is also needed.

I tried add 
```
RUSTFLAGS = "--cfg tokio_unstable";
RUSTC_BOOTSTRAP = 1;
```

and there seems no change.

**Comment by linsui** [comment]
```nix
  RUSTC_BOOTSTRAP = true;
  RUSTFLAGS = "--cfg tokio_unstable";
```

and `rev = "161840e3ff40e82701aa95fd168e929744e35cdd";` work for me.

**Comment by oluceps** [comment]
> ```nix
>   RUSTC_BOOTSTRAP = true;
>   RUSTFLAGS = "--cfg tokio_unstable";
> ```
> 
> and `rev = "161840e3ff40e82701aa95fd168e929744e35cdd";` work for me.
> 
> Full file:
> 
> ```nix
> { fetchFromGitHub
> , rustPlatform
> , lib
> , ipset
> }:
> 
> rustPlatform.buildRustPackage rec{
>   pname = "trojan-rs";
>   version = "0.16.0-unstable-2024-02-18";
> 
>   src = fetchFromGitHub {
>     owner = "lazytiger";
>     repo = pname;
>     rev = "161840e3ff40e82701aa95fd168e929744e35cdd";
>     hash = "sha256-9c90WDP1TuNZwA3jOqwAqpSJzbFpPMtNUYUXSncJgKg=";
>   };
> 
>   cargoHash = "sha256-TDUi7MXp9fw8CoxhqQ9GBdIOOilsnn8tU2rW/QuT+jc=";
> 
>   buildInputs = [ ipset ];
> 
>   RUSTC_BOOTSTRAP = true;
>   RUSTFLAGS = "--cfg tokio_unstable";
> 
>   meta = with lib; {
>     homepage = "https://github.com/lazytiger/trojan-rs";
>     description = "Trojan server and proxy programs written in Rust";
>     license = licenses.mit;
>     mainProgram = "trojan";
>     maintainers = with maintainers; [ oluceps ];
>   };
> }
> ```

That works, thank you! Would you like to be the co-maintainer?

**Comment by linsui** [comment]
> Would you like to be the co-maintainer?

No. :)

**Review by linsui — APPROVED** [review]

**Review by linsui — APPROVED** [review]

File: pkgs/by-name/tr/trojan-rs/package.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{ fetchFromGitHub
+, rustPlatform
```
Comment (FliegendeWurst) [inline pkgs/by-name/tr/trojan-rs/package.nix:2]: Please format with the new formatter.

File: pkgs/by-name/tr/trojan-rs/package.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    rev = "161840e3ff40e82701aa95fd168e929744e35cdd";
+    hash = "sha256-9c90WDP1TuNZwA3jOqwAqpSJzbFpPMtNUYUXSncJgKg=";
+  };
+
+  cargoHash = "sha256-ZWVkdU+e/X/io1Sb2epugg0Qv5vgrsnB9UuuT0VpKi4=";
+
+  buildInputs = [ ipset ];
+
+  RUSTC_BOOTSTRAP = true;
+  RUSTFLAGS = "--cfg tokio_unstable";
```
Comment (FliegendeWurst) [inline pkgs/by-name/tr/trojan-rs/package.nix:23]:
```suggestion
  env.RUSTC_BOOTSTRAP = true;
  env.RUSTFLAGS = "--cfg tokio_unstable";
```

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Review by FliegendeWurst — APPROVED** [review]

**Review by fpletz — APPROVED** [review]

**Pull request merged**